File: apps/scanner/app/src/main/java/com/airferry/app/scan/Af2LedgerStore.kt
package com.airferry.app.scan

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.RandomAccessFile
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.UUID
import org.json.JSONObject

/**
 * Crash-safe §12 resume ledger — the journal twin of [ChunkSpillStore]'s
 * `.partial` file.
 *
 * Format: one JSONL journal per transfer, `af2-<tid>.ledger.jsonl`.
 * - Line 1 (header): `{"v":1,"tid":…,"rid":…,"root":…,"crs":…}` — `rid`
 *   identifies this receive attempt (legacy headers omit it); written once,
 *   atomically (temp file + fsync + rename) before the first chunk commit.
 * - Each later line: `{"c":<index>}` (chunk committed after spill+fsync) or
 *   `{"i":<index>}` (chunk invalidated after a re-verification failure).
 *
 * Per-line append + fsync keeps every interleaving crash-safe: only a torn
 * final line is skipped; corruption before the tail rejects the candidate,
 * so the ledger never reports MORE than what hit the disk. A commit line is only appended after
 * the chunk bytes were pwrite + fsync'd into the spill — the §12 ordering
 * rule "账本完成 ⇒ 数据已落盘" holds by construction.
 *
 * Only touched from the ingest thread (the decode pool serializes ingest)
 * and the recovery path that runs under the same lock.
 */
class Af2LedgerStore private constructor(private val path: File) {

    var transferIdHex: String = ""
        private set
    /** One receive attempt, persisted across process restart. Unlike Transfer
     * ID this changes when the same content is intentionally received again. */
    var recoveryId: String = ""
        private set
    var chunkRawSize: Int = 0
        private set
    /** ROOT frame bytes (hex at rest) for the §12 resume() call. */
    var rootFrameBytes: ByteArray = ByteArray(0)
        private set
    private val completed = sortedSetOf<Int>()
    private var headerDurable = false

    val completedIndices: IntArray get() = completed.toIntArray()

    /** Load an existing ledger for `tid`, or null when none exists. */
    fun reload(): Boolean {
        completed.clear()
        transferIdHex = ""
        recoveryId = ""
        chunkRawSize = 0
        rootFrameBytes = ByteArray(0)
        headerDurable = false
        if (!path.isFile) return false
        if (path.length() > MAX_LEDGER_BYTES) return false
        val text = try {
            path.readText()
        } catch (_: Exception) {
            return false
        }
        val originalByteLength = text.toByteArray(Charsets.UTF_8).size.toLong()
        val tailTerminated = text.endsWith('\n')
        val durableText = if (tailTerminated) {
            text
        } else {
            val lastNewline = text.lastIndexOf('\n')
            if (lastNewline < 0) return false // even the header lacks its commit delimiter
            text.substring(0, lastNewline + 1)
        }
        val splitLines = durableText.split('\n')
        val lines = if (splitLines.lastOrNull()?.isEmpty() == true) {
            splitLines.dropLast(1)
        } else {
            splitLines
        }
        // A blank physical record is corruption; filtering it out could hide
        // a lost invalidation. Only an unterminated final JSON fragment can be
        // the product of a crash-torn append and may be skipped.
        if (lines.isEmpty() || lines.any { it.isBlank() }) return false
        val header = try {
            JSONObject(lines.first())
        } catch (_: Exception) {
            return false
        }
        val fileTid = transferIdFromLedgerName(path.name) ?: return false
        val parsedTid = header.optString("tid", "")
        val hasRecoveryId = header.has("rid")
        val parsedRecoveryId = if (hasRecoveryId) header.optString("rid", "") else parsedTid
        val parsedVersion = strictInt(header, "v")
        val parsedChunkRawSize = strictInt(header, "crs") ?: 0
        val rootHex = header.optString("root", "")
        val parsedRoot = if (rootHex.length <= MAX_ROOT_FRAME_HEX_CHARS) {
            hexToBytes(rootHex)
        } else {
            ByteArray(0)
        }
        if (
            header.length() != (if (hasRecoveryId) 5 else 4) ||
            header.opt("tid") !is String ||
            header.opt("root") !is String ||
            (hasRecoveryId && header.opt("rid") !is String) ||
            parsedVersion != 1 ||
            parsedTid != fileTid ||
            !parsedRecoveryId.matches(SAFE_RECOVERY_ID) ||
            parsedChunkRawSize !in LEGAL_CHUNK_RAW_SIZES ||
            parsedRoot.isEmpty()
        ) return false

        val parsedCompleted = sortedSetOf<Int>()
        for (position in 1 until lines.size) {
            val o = try {
                JSONObject(lines[position])
            } catch (_: Exception) {
                return false
            }
            val key = when {
                o.length() == 1 && o.has("c") -> "c"
                o.length() == 1 && o.has("i") -> "i"
                else -> return false
            }
            val index = strictInt(o, key) ?: return false
            if (index !in 0 until MAX_CHUNK_COUNT) return false
            if (key == "c") parsedCompleted.add(index) else parsedCompleted.remove(index)
        }
        if (!tailTerminated) {
            // Never append after a crash fragment: `fragment + next JSON`
            // would turn recoverable work into permanent middle corruption.
            // A record is committed only by its trailing newline, so even a
            // syntactically-complete unterminated JSON object is discarded.
            val durableByteLength = durableText.toByteArray(Charsets.UTF_8).size.toLong()
            try {
                RandomAccessFile(path, "rw").use { file ->
                    if (file.length() != originalByteLength) return false
                    file.setLength(durableByteLength)
                    file.fd.sync()
                }
            } catch (_: Exception) {
                return false
            }
        }
        transferIdHex = parsedTid
        recoveryId = parsedRecoveryId
        chunkRawSize = parsedChunkRawSize
        rootFrameBytes = parsedRoot
        completed.addAll(parsedCompleted)
        headerDurable = true
        return true
    }

    /** Append one commit event (after the chunk was spilled + fsync'd). */
    @Throws(IOException::class)
    fun commit(index: Int) {
        require(index in 0 until MAX_CHUNK_COUNT) { "invalid AF2 chunk index: $index" }
        if (!headerDurable) throw IOException("AF2 ledger header is not durable")
        appendLine(JSONObject().put("c", index))
        completed.add(index)
    }

    /** Append one invalidate event (after a spill re-verification failure). */
    @Throws(IOException::class)
    fun invalidate(index: Int) {
        require(index in 0 until MAX_CHUNK_COUNT) { "invalid AF2 chunk index: $index" }
        if (!headerDurable) throw IOException("AF2 ledger header is not durable")
        appendLine(JSONObject().put("i", index))
        completed.remove(index)
    }

    @Throws(IOException::class)
    private fun appendLine(o: JSONObject) {
        FileOutputStream(path, true).use { fos ->
            fos.write((o.toString() + "\n").toByteArray())
            fos.fd.sync()
        }
    }

    /** Delete the journal (transfer finished / relocked away / abandoned). */
    fun discard() {
        path.delete()
        headerDurable = false
    }

    data class PendingTransfer(
        val transferIdHex: String,
        val chunkRawSize: Int,
        val completedCount: Int,
        val diskBytes: Long,
        val lastModified: Long,
    )

    companion object {
        private const val MAX_CHUNK_COUNT = 131_072
        private const val MAX_LEDGER_BYTES = 32L * 1024 * 1024
        private const val MAX_ROOT_FRAME_BYTES = 26 + 2400 + 4
        private const val MAX_ROOT_FRAME_HEX_CHARS = MAX_ROOT_FRAME_BYTES * 2
        private val LEGAL_CHUNK_RAW_SIZES = setOf(1, 2, 4, 8, 16, 32).mapTo(mutableSetOf()) {
            it * 1024 * 1024
        }
        internal val SAFE_TRANSFER_ID = Regex("^[A-Za-z0-9_-]{1,64}$")
        private val SAFE_RECOVERY_ID = Regex("^[A-Za-z0-9_-]{1,64}$")

        private fun transferIdFromLedgerName(name: String): String? {
            val prefix = "af2-"
            val suffix = ".ledger.jsonl"
            if (!name.startsWith(prefix) || !name.endsWith(suffix)) return null
            return name.removePrefix(prefix).removeSuffix(suffix).takeIf { it.matches(SAFE_TRANSFER_ID) }
        }

        /** List all uncompleted/partial transfer ledgers in `dir`. */
        fun listPendingTransfers(dir: File): List<PendingTransfer> {
            val candidates = dir.listFiles { f -> transferIdFromLedgerName(f.name) != null }
                ?: return emptyList()
            val list = mutableListOf<PendingTransfer>()
            for (f in candidates) {
                val store = Af2LedgerStore(f)
                if (store.reload()) {
                    val tid = store.transferIdHex
                    val spill = File(dir, "af2-$tid.partial")
                    val spillBytes = if (spill.isFile) spill.length() else 0L
                    list.add(
                        PendingTransfer(
                            transferIdHex = tid,
                            chunkRawSize = store.chunkRawSize,
                            completedCount = store.completedIndices.size,
                            diskBytes = f.length() + spillBytes,
                            lastModified = f.lastModified()
                        )
                    )
                }
            }
            return list.sortedByDescending { it.lastModified }
        }

        /** Discard all pending journals and spill files in `dir`. */
        fun discardAllPending(dir: File) {
            dir.listFiles { f ->
                f.name.startsWith("af2-") && (
                    f.name.endsWith(".ledger.jsonl") ||
                    f.name.endsWith(".partial") ||
                    f.name.endsWith(".tmp")
                )
            }?.forEach { it.delete() }
        }

        /** Resume source: newest valid ledger in `dir` (by mtime), or null. */
        fun loadMostRecent(dir: File): Af2LedgerStore? {
            val candidates = dir.listFiles { f -> transferIdFromLedgerName(f.name) != null }
                ?: return null
            for (candidate in candidates.sortedByDescending { it.lastModified() }) {
                try {
                    val store = Af2LedgerStore(candidate)
                    if (store.reload()) return store
                } catch (_: Exception) {
                    // One corrupt candidate must not hide older valid work.
                }
            }
            return null
        }

        /** Remove unrecoverable spill files that have no valid resume journal. */
        fun sweepOrphanPartials(dir: File) {
            val validTids = mutableSetOf<String>()
            dir.listFiles { f -> transferIdFromLedgerName(f.name) != null }?.forEach { file ->
                val store = Af2LedgerStore(file)
                if (store.reload()) {
                    validTids.add(store.transferIdHex)
                } else {
                    file.delete()
                }
            }
            dir.listFiles { f -> f.name.startsWith("af2-") && f.name.endsWith(".partial") }
                ?.forEach { partial ->
                    val tid = partial.name.removePrefix("af2-").removeSuffix(".partial")
                    if (tid !in validTids) partial.delete()
                }
        }

        /** Create + write the header for a fresh transfer's journal. */
        fun create(
            dir: File,
            transferIdHex: String,
            chunkRawSize: Int,
            rootFrameBytes: ByteArray
        ): Af2LedgerStore {
            require(transferIdHex.matches(SAFE_TRANSFER_ID)) { "invalid AF2 transfer id" }
            require(chunkRawSize in LEGAL_CHUNK_RAW_SIZES) { "invalid AF2 chunk size" }
            require(rootFrameBytes.size in 1..MAX_ROOT_FRAME_BYTES) {
                "invalid AF2 ROOT frame"
            }
            if ((!dir.exists() && !dir.mkdirs()) || !dir.isDirectory) {
                throw IOException("AF2 ledger directory is unavailable")
            }
            val path = File(dir, "af2-$transferIdHex.ledger.jsonl")
            val recoveryId = UUID.randomUUID().toString()
            val header = JSONObject()
                .put("v", 1)
                .put("tid", transferIdHex)
                .put("rid", recoveryId)
                .put("crs", chunkRawSize)
                .put("root", bytesToHex(rootFrameBytes))
            // Atomic header: temp + fsync + rename so a crash mid-create
            // never leaves a headerless journal that a later commit would
            // append to. Keep an existing same-transfer ledger intact until
            // the replacement header itself is durable: a disk-full failure
            // during relock must not destroy otherwise resumable work.
            val tmp = File(dir, "${path.name}.${UUID.randomUUID()}.tmp")
            try {
                FileOutputStream(tmp).use { fos ->
                    fos.write((header.toString() + "\n").toByteArray())
                    fos.fd.sync()
                }
                try {
                    Files.move(
                        tmp.toPath(),
                        path.toPath(),
                        StandardCopyOption.ATOMIC_MOVE,
                        StandardCopyOption.REPLACE_EXISTING,
                    )
                } catch (_: AtomicMoveNotSupportedException) {
                    // Both paths live in the same cache directory. Even when
                    // the provider lacks ATOMIC_MOVE, a replace-rename occurs
                    // only after the new header has been flushed.
                    Files.move(
                        tmp.toPath(),
                        path.toPath(),
                        StandardCopyOption.REPLACE_EXISTING,
                    )
                }
            } catch (e: Exception) {
                throw IOException("AF2 ledger header write failed", e)
            } finally {
                tmp.delete()
            }
            return Af2LedgerStore(path).apply {
                this.transferIdHex = transferIdHex
                this.recoveryId = recoveryId
                this.chunkRawSize = chunkRawSize
                this.rootFrameBytes = rootFrameBytes.copyOf()
                this.headerDurable = true
            }
        }

        /** Drop the journal (and spill) of `tid` — completion cleanup. */
        fun discardFor(dir: File, transferIdHex: String) {
            File(dir, "af2-$transferIdHex.ledger.jsonl").delete()
        }

        private fun bytesToHex(b: ByteArray): String =
            b.joinToString("") { "%02x".format(it) }

        private fun hexToBytes(s: String): ByteArray {
            if (s.length % 2 != 0) return ByteArray(0)
            return ByteArray(s.length / 2) { i ->
                val hi = Character.digit(s[i * 2], 16)
                val lo = Character.digit(s[i * 2 + 1], 16)
                if (hi < 0 || lo < 0) return ByteArray(0)
                ((hi shl 4) + lo).toByte()
            }
        }

        private fun strictInt(o: JSONObject, key: String): Int? = when (val value = o.opt(key)) {
            is Int -> value
            is Long -> value.takeIf {
                it in Int.MIN_VALUE.toLong()..Int.MAX_VALUE.toLong()
            }?.toInt()
            else -> null
        }
    }
}

File: apps/windows/AirFerry.Windows/Scan/Af2LedgerStore.cs
using System.IO;
using System.Text;
using System.Text.Json;

namespace AirFerry.Windows.Scan;

/// <summary>
/// Crash-safe §12 resume ledger — the journal twin of <see cref="ChunkSpillStore"/>'s
/// <c>.partial</c> file. JSONL, one file per transfer (<c>af2-&lt;tid&gt;.ledger.jsonl</c>):
/// <para>
/// Line 1 (header): <c>{"v":1,"tid":…,"rid":…,"root":…,"crs":…}</c> — <c>rid</c>
/// identifies this receive attempt (legacy headers omit it); written atomically
/// (temp + flush + rename) before the first chunk commit. Each later line is
/// <c>{"c":i}</c> (chunk committed after its bytes were pwrite+fsync'd into the
/// spill) or <c>{"i":i}</c> (chunk invalidated after a re-verification failure).
/// Only a torn final line is skipped; earlier corruption rejects the candidate,
/// so the journal never reports more than what reached the disk.
/// </para>
/// <para>
/// Only touched from the pool's serialized ingest callback (under
/// <c>IngestLock</c>) and the recovery core that runs under the same lock.
/// </para>
/// </summary>
public sealed class Af2LedgerStore
{
    private const int MaxChunkCount = 131_072;
    private const long MaxLedgerBytes = 32L * 1024 * 1024;
    private const int MaxRootFrameBytes = 26 + 2400 + 4;
    private static readonly HashSet<int> LegalChunkRawSizes =
        [1 << 20, 2 << 20, 4 << 20, 8 << 20, 16 << 20, 32 << 20];

    private readonly string _path;
    public string TransferIdHex { get; private set; } = "";
    /// <summary>Identity of one receive attempt; survives restart but changes
    /// when the same Transfer is intentionally received again.</summary>
    public string RecoveryId { get; private set; } = "";
    public int ChunkRawSize { get; private set; }
    public byte[] RootFrameBytes { get; private set; } = Array.Empty<byte>();
    public SortedSet<int> Completed { get; } = new();

    private Af2LedgerStore(string path)
    {
        _path = path;
    }

    /// <summary>True once the header line is durably on disk (set by Create).</summary>
    private bool _headerDurable;

    public int[] CompletedIndices => Completed.ToArray();

    /// <summary>Parse (or re-parse) the journal. True when a valid header exists.</summary>
    public bool Reload()
    {
        Completed.Clear();
        TransferIdHex = "";
        RecoveryId = "";
        ChunkRawSize = 0;
        RootFrameBytes = [];
        _headerDurable = false;
        if (!File.Exists(_path))
        {
            return false;
        }
        long originalByteLength = new FileInfo(_path).Length;
        if (originalByteLength > MaxLedgerBytes) return false;
        string text;
        try
        {
            text = File.ReadAllText(_path);
        }
        catch (IOException)
        {
            return false;
        }
        bool tailTerminated = text.EndsWith('\n');
        string durableText;
        if (tailTerminated)
        {
            durableText = text;
        }
        else
        {
            int lastNewline = text.LastIndexOf('\n');
            if (lastNewline < 0) return false;
            durableText = text[..(lastNewline + 1)];
        }
        var records = durableText.Split('\n').ToList();
        if (records.Count > 0 && records[^1].Length == 0)
            records.RemoveAt(records.Count - 1);
        // Blank physical records are corruption, not ignorable whitespace.
        // Only a malformed final fragment without its newline can be a torn
        // append; accepting a newline-terminated bad record can hide an
        // invalidation and resurrect corrupt spill bytes.
        if (records.Count == 0 || records.Any(string.IsNullOrWhiteSpace)) return false;

        JsonElement header;
        try
        {
            using JsonDocument doc = JsonDocument.Parse(records[0]);
            header = doc.RootElement.Clone();
        }
        catch (JsonException)
        {
            return false;
        }
        if (header.ValueKind != JsonValueKind.Object) return false;
        string? fileTid = TransferIdFromLedgerName(Path.GetFileName(_path));
        JsonProperty[] headerProperties = header.EnumerateObject().ToArray();
        string parsedTid = header.TryGetProperty("tid", out JsonElement tid) &&
            tid.ValueKind == JsonValueKind.String
            ? tid.GetString() ?? "" : "";
        bool hasRecoveryId = header.TryGetProperty("rid", out JsonElement rid);
        string parsedRecoveryId = hasRecoveryId && rid.ValueKind == JsonValueKind.String
            ? rid.GetString() ?? ""
            : hasRecoveryId ? "" : parsedTid;
        int parsedChunkRawSize = header.TryGetProperty("crs", out JsonElement crs) &&
            crs.TryGetInt32(out int crsValue) ? crsValue : 0;
        string rootHex = header.TryGetProperty("root", out JsonElement root) &&
            root.ValueKind == JsonValueKind.String
            ? root.GetString() ?? "" : "";
        byte[] parsedRoot = rootHex.Length <= MaxRootFrameBytes * 2
            ? HexToBytes(rootHex) : [];
        if (headerProperties.Length != (hasRecoveryId ? 5 : 4) ||
            !header.TryGetProperty("v", out JsonElement version) ||
            !version.TryGetInt32(out int versionValue) || versionValue != 1 ||
            fileTid is null || !string.Equals(parsedTid, fileTid, StringComparison.Ordinal) ||
            !IsSafeRecoveryId(parsedRecoveryId) ||
            !LegalChunkRawSizes.Contains(parsedChunkRawSize) ||
            parsedRoot.Length == 0)
        {
            return false;
        }

        var parsedCompleted = new SortedSet<int>();
        for (int position = 1; position < records.Count; position++)
        {
            JsonElement record;
            try
            {
                using JsonDocument doc = JsonDocument.Parse(records[position]);
                record = doc.RootElement.Clone();
            }
            catch (JsonException)
            {
                return false;
            }
            JsonProperty[] properties = record.ValueKind == JsonValueKind.Object
                ? record.EnumerateObject().ToArray() : [];
            if (properties.Length != 1 ||
                !properties[0].Value.TryGetInt32(out int index) ||
                index < 0 || index >= MaxChunkCount)
            {
                return false;
            }
            if (properties[0].NameEquals("c"))
            {
                parsedCompleted.Add(index);
            }
            else if (properties[0].NameEquals("i"))
            {
                parsedCompleted.Remove(index);
            }
            else return false;
        }
        if (!tailTerminated)
        {
            // A newline is the commit delimiter. Drop even a complete JSON
            // object when that delimiter was torn, then fsync the repaired
            // prefix before this resumed store is allowed to append again.
            try
            {
                long durableByteLength = Encoding.UTF8.GetByteCount(durableText);
                using var stream = new FileStream(
                    _path, FileMode.Open, FileAccess.Write, FileShare.None);
                if (stream.Length != originalByteLength) return false;
                stream.SetLength(durableByteLength);
                stream.Flush(flushToDisk: true);
            }
            catch (Exception)
            {
                return false;
            }
        }
        // A journal that reloads with a valid header is durable by definition —
        // resumed transfers keep appending to it.
        TransferIdHex = parsedTid;
        RecoveryId = parsedRecoveryId;
        ChunkRawSize = parsedChunkRawSize;
        RootFrameBytes = parsedRoot;
        Completed.UnionWith(parsedCompleted);
        _headerDurable = true;
        return true;
    }

    /// <summary>Append one commit event (after the chunk was spilled + flushed).</summary>
    public void Commit(int index)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index);
        if (index >= MaxChunkCount) throw new ArgumentOutOfRangeException(nameof(index));
        if (!_headerDurable)
            throw new IOException("AF2 ledger header is not durable");
        AppendLine($"{{\"c\":{index}}}");
        Completed.Add(index);
    }

    /// <summary>Append one invalidate event (after a re-verification failure).</summary>
    public void Invalidate(int index)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index);
        if (index >= MaxChunkCount) throw new ArgumentOutOfRangeException(nameof(index));
        if (!_headerDurable)
            throw new IOException("AF2 ledger header is not durable");
        AppendLine($"{{\"i\":{index}}}");
        Completed.Remove(index);
    }

    private void AppendLine(string json)
    {
        using var fs = new FileStream(
            _path, FileMode.Append, FileAccess.Write, FileShare.None);
        byte[] bytes = Encoding.UTF8.GetBytes(json + "\n");
        fs.Write(bytes, 0, bytes.Length);
        fs.Flush(flushToDisk: true);
    }

    /// <summary>Delete the journal (transfer finished / relocked away / abandoned).</summary>
    public void Discard()
    {
        try { File.Delete(_path); } catch (IOException) { }
        _headerDurable = false;
    }

    /// <summary>Delete this journal and its same-transfer spill backing.</summary>
    public void DiscardTransferFiles()
    {
        Discard();
        const string suffix = ".ledger.jsonl";
        if (!_path.EndsWith(suffix, StringComparison.Ordinal)) return;
        string spill = _path[..^suffix.Length] + ".partial";
        try { File.Delete(spill); } catch (IOException) { }
    }

    public record PendingTransfer(
        string TransferIdHex,
        int ChunkRawSize,
        int CompletedCount,
        long DiskBytes,
        DateTime LastModified);

    /// <summary>List all pending/uncompleted transfer ledgers in <paramref name="dir"/>.</summary>
    public static IReadOnlyList<PendingTransfer> ListPendingTransfers(string dir)
    {
        if (!Directory.Exists(dir)) return [];
        try
        {
            var list = new List<PendingTransfer>();
            foreach (var file in Directory.EnumerateFiles(dir, "af2-*.ledger.jsonl"))
            {
                if (TransferIdFromLedgerName(Path.GetFileName(file)) is null) continue;
                var store = new Af2LedgerStore(file);
                if (store.Reload())
                {
                    string tid = store.TransferIdHex;
                    string spill = Path.Combine(dir, $"af2-{tid}.partial");
                    long spillBytes = File.Exists(spill) ? new FileInfo(spill).Length : 0L;
                    var fi = new FileInfo(file);
                    list.Add(new PendingTransfer(
                        tid,
                        store.ChunkRawSize,
                        store.CompletedIndices.Length,
                        fi.Length + spillBytes,
                        fi.LastWriteTime));
                }
            }
            return list.OrderByDescending(p => p.LastModified).ToList();
        }
        catch (Exception)
        {
            return [];
        }
    }

    /// <summary>Discard all pending journals, spills, and temp files in <paramref name="dir"/>.</summary>
    public static void DiscardAllPending(string dir)
    {
        if (!Directory.Exists(dir)) return;
        try
        {
            foreach (var f in Directory.EnumerateFiles(dir, "af2-*"))
            {
                if (f.EndsWith(".ledger.jsonl") || f.EndsWith(".partial") || f.EndsWith(".tmp"))
                {
                    try { File.Delete(f); } catch { }
                }
            }
        }
        catch { }
    }

    /// <summary>Resume source: the newest valid journal in <paramref name="dir"/>.</summary>
    public static Af2LedgerStore? LoadMostRecent(string dir)
    {
        try
        {
            if (!Directory.Exists(dir))
            {
                return null;
            }
            foreach (FileInfo candidate in new DirectoryInfo(dir)
                         .EnumerateFiles("af2-*.ledger.jsonl")
                         .Where(f => TransferIdFromLedgerName(f.Name) is not null)
                         .OrderByDescending(f => f.LastWriteTimeUtc))
            {
                try
                {
                    var store = new Af2LedgerStore(candidate.FullName);
                    if (store.Reload()) return store;
                }
                catch
                {
                    // One malformed candidate must not hide older valid work.
                }
            }
            return null;
        }
        catch (Exception)
        {
            return null;
        }
    }

    /// <summary>Remove spill files that have no parseable resume journal.</summary>
    public static void SweepOrphanPartials(string dir)
    {
        if (!Directory.Exists(dir)) return;
        try
        {
            var validTids = new HashSet<string>(StringComparer.Ordinal);
            foreach (string journal in Directory.EnumerateFiles(dir, "af2-*.ledger.jsonl"))
            {
                if (TransferIdFromLedgerName(Path.GetFileName(journal)) is null) continue;
                var store = new Af2LedgerStore(journal);
                if (store.Reload())
                {
                    validTids.Add(store.TransferIdHex);
                }
                else
                {
                    try { File.Delete(journal); } catch { }
                }
            }
            foreach (string partial in Directory.EnumerateFiles(dir, "af2-*.partial"))
            {
                string name = Path.GetFileName(partial);
                string tid = name.Substring(4, name.Length - 4 - ".partial".Length);
                if (!validTids.Contains(tid))
                {
                    try { File.Delete(partial); } catch { }
                }
            }
        }
        catch { }
    }

    /// <summary>Create + atomically write the header for a fresh transfer's journal.</summary>
    public static Af2LedgerStore Create(
        string dir, string transferIdHex, int chunkRawSize, byte[] rootFrameBytes)
    {
        if (!IsSafeTransferId(transferIdHex)) throw new ArgumentException("Invalid AF2 transfer id", nameof(transferIdHex));
        if (!LegalChunkRawSizes.Contains(chunkRawSize)) throw new ArgumentOutOfRangeException(nameof(chunkRawSize));
        if (rootFrameBytes.Length == 0 || rootFrameBytes.Length > MaxRootFrameBytes)
            throw new ArgumentException("Invalid AF2 ROOT frame", nameof(rootFrameBytes));
        string path = Path.Combine(dir, $"af2-{transferIdHex}.ledger.jsonl");
        string recoveryId = Guid.NewGuid().ToString("N");
        string header = JsonSerializer.Serialize(new
        {
            v = 1,
            tid = transferIdHex,
            rid = recoveryId,
            crs = chunkRawSize,
            root = BytesToHex(rootFrameBytes),
        });
        // A unique temp avoids stale/concurrent temp collisions. Keep any
        // existing same-transfer ledger until the new header is flushed and
        // ready to replace it, so a failed relock cannot erase resumable work.
        string tmp = $"{path}.{Guid.NewGuid():N}.tmp";
        try
        {
            Directory.CreateDirectory(dir);
            byte[] bytes = Encoding.UTF8.GetBytes(header + "\n");
            using (var fs = new FileStream(tmp, FileMode.Create, FileAccess.Write, FileShare.None))
            {
                fs.Write(bytes, 0, bytes.Length);
                fs.Flush(flushToDisk: true);
            }
            File.Move(tmp, path, overwrite: true);
        }
        catch (Exception ex)
        {
            try { File.Delete(tmp); } catch { }
            throw new IOException("AF2 ledger header write failed", ex);
        }
        return new Af2LedgerStore(path)
        {
            TransferIdHex = transferIdHex,
            RecoveryId = recoveryId,
            ChunkRawSize = chunkRawSize,
            RootFrameBytes = rootFrameBytes.ToArray(),
            _headerDurable = true,
        };
    }

    private static string BytesToHex(byte[] b)
    {
        var sb = new StringBuilder(b.Length * 2);
        foreach (byte x in b)
        {
            sb.Append(x.ToString("x2"));
        }
        return sb.ToString();
    }

    private static byte[] HexToBytes(string s)
    {
        if (s.Length == 0 || s.Length % 2 != 0)
        {
            return Array.Empty<byte>();
        }
        try { return Convert.FromHexString(s); }
        catch (FormatException) { return Array.Empty<byte>(); }
    }

    private static string? TransferIdFromLedgerName(string name)
    {
        const string prefix = "af2-";
        const string suffix = ".ledger.jsonl";
        if (!name.StartsWith(prefix, StringComparison.Ordinal) ||
            !name.EndsWith(suffix, StringComparison.Ordinal)) return null;
        string id = name[prefix.Length..^suffix.Length];
        return IsSafeTransferId(id) ? id : null;
    }

    internal static bool IsSafeTransferId(string id) =>
        id.Length is >= 1 and <= 64 && id.All(c =>
            char.IsAsciiLetterOrDigit(c) || c is '_' or '-');

    private static bool IsSafeRecoveryId(string id) => IsSafeTransferId(id);
}
